[최지수] Step4 리뷰 요청합니다~

jisu1211 opened the pull request:
안녕하세요, 우빈님~
3단계 피드백 반영과 4단계 구현 완료하여 코드 리뷰 요청합니다!
저번 피드백에서 자바 컨벤션에 대한 내용을 말씀해주셔서,
전체적으로 열 라인을 100라인으로 맞추고 논리적인 단위로 코드 줄바꿈도 해봤습니다~
이번에도 피드백 잘 부탁드립니다 ㅎㅎ
감사합니다😊

jisu1211 committed:
docs: README 3단계 수정 / 4단계 추가

jisu1211 committed:
fix: 3단계 피드백 반영
- 불필요한 static 선언 제거
- Stream 사용시 메서드 참조로 간결하게 수정
- 자바 컨벤션을 참조해 전체적인 열 라인(100 기준), 줄바꿈 수정
- this 명령어 추가
- LottoGames 패키지 변경

feat: 4단계 로또(2등) 기능 추가
- LottoGames
  - 보너스 번호를 입력받는다.
  - 1-5등 당첨 결과와 수익률을 출력한다.
- Lotto
  - 지난 주 당첨 번호를 입력받아 당첨 상태(1-5등/낙첨)를 가진다.
  - LottoStatus : 2등 당첨 상태와 보너스 넘버가 일치할 때 2등 당첨을 결정하는 로직 추가
- WinningLotto
  - 보너스 번호를 가진다.
    - 지난 주 당첨번호에 포함된 번호가 보너스 번호로 입력되면 IllegalArgumentException 발생
    - 보너스 번호가 사용자가 구매한 로또에 포함되었는지 확인한다.

- test: 변경된 로직에 따른 테스트 코드 추가/변경

limwoobin commented on README.md:
```diff
@@ -78,4 +80,37 @@
\ No newline at end of file
+총 수익률은 0.35입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)
+
+## 4단계 : 로또(2등)
+[클래스 구현]
+* LottoGames
+  * 보너스 번호를 추가로 입력받을 수 있다.
+  * 1-5등 당첨 결과와 수익률을 출력한다.
+* Lotto
+  * 지난 주 당첨 번호를 입력받아 당첨 상태(1-5등/낙첨)를 가진다.
+  * LottoStatus: 당첨 상태 클래스
+    * 1등: 6개 숫자 일치
+    * 2등: 5개 숫자 일치 + 보너스볼 일치
+    * 3등: 5개 숫자 일치
+    * 4등: 4개 숫자 일치
+    * 5등: 3개 숫자 일치
+* WinningLotto
+  * 보너스 번호를 가진다.
+    * 지난 주 당첨번호에 포함된 번호가 보너스 번호로 입력되면 IllegalArgumentException 발생
+    * 보너스 번호가 사용자가 구매한 로또에 포함되었는지 확인한다.
+
+[입출력 예시]\
+지난 주 당첨 번호를 입력해 주세요.\
+1, 2, 3, 4, 5, 6\
+보너스 볼을 입력해 주세요.\
+7\
+\
+당첨 통계\
+---------\
+3개 일치 (5000원)- 1개\
+4개 일치 (50000원)- 0개\
+5개 일치 (1500000원)- 0개\
+5개 일치, 보너스 볼 일치(30000000원) - 0개\
+6개 일치 (2000000000원)- 0개\
+총 수익률은 0.35입니다.(기준이 1이기 때문에 결과적으로 손해라는 의미임)\
```
요구사항 정리 👍

jisu1211 replied:
커밋 단위를 미리 나눴어야 했는데 보기 힘드셨을까봐 죄송하네요 ㅠㅠ

limwoobin commented on src/main/java/study/lotto/domain/LottoStatus.java:
```diff
@@ -7,10 +7,13 @@
+    FOURTH_PLACE(4, 50_000),
+    THIRD_PLACE(5, 1_500_000),
+    SECOND_PLACE(5, 30_000_000),
+    FIRST_PLACE(6, 2_000_000_000);
```
천단위로 구분을 해주셨군요 
전보다 가독성이 좋습니다!

limwoobin commented on src/main/java/study/lotto/domain/Lotto.java:
```diff
@@ -7,18 +7,15 @@
     }
 
-    private static Set<LottoNumber> toSet(List<Integer> numbersFromStore) {
+    private Set<LottoNumber> toSet(List<Integer> numbersFromStore) {
```
`toSet` 이라는 이름보다는 로또번호를 생성하거나 가져온다라는 의미를
표현할 수 있는 이름은 어떨까요?

jisu1211 replied:
네 고민해보겠습니다!

limwoobin commented on src/main/java/study/lotto/domain/LottoNumber.java:
```diff
@@ -15,7 +15,7 @@ public class LottoNumber implements Comparable<LottoNumber> {
 
     static {
         IntStream.rangeClosed(LOTTO_MIN_NUM, LOTTO_MAX_NUM)
-                .forEach((num) -> lottoNumbers.put(num, new LottoNumber(num)));
+            .forEach((num) -> lottoNumbers.put(num, new LottoNumber(num)));
     }
```
`LottoNumber`도 1부터 45까지 캐싱을 해주셨군요
만약 그렇다면 이 lottoNumbers를 갖고있는 주체가 LottoNumber가 맞을까요??

jisu1211 replied:
그 부분까지는 생각을 못해봤는데 한 번 고민해보겠습니다 ㅎㅎ 감사합니다!

limwoobin commented on src/main/java/study/lotto/domain/Order.java:
```diff
@@ -14,9 +14,11 @@ public Order(String totalAmount) {
             return totalAmountConverted;
         }
+
         throw new IllegalArgumentException("[ERROR] You must purchase at least one lotto.");
```
예외에 대해 스트링으로 바로 리턴하는 부분이 상당히 보이네요 ㅎㅎ
이런 메시지는 별도의 message constant 객체를 이용하거나 enum을 이용하면
보다 깔끔하고 관리하기 쉽게 사용할 수 있습니다!

jisu1211 replied:
에러 메시지 분류해보겠습니다~!

limwoobin commented on src/main/java/study/lotto/domain/Store.java:
```diff
@@ -28,16 +30,20 @@ private static List<Integer> getRangeOfLottoNumbers() {
+                .stream()
+                .map(LottoNumber::of)
                 .collect(Collectors.toSet());
     }
```
buy라는 이름은 상위 메소드에서 이미 사용하고 있습니다
그렇다면 이 메소드는 구매한 로또를 생성하거나 가져오는걸로 보여지는데 
알맞은 다른 메소드명이 있지않을까요??

limwoobin commented on src/main/java/study/lotto/domain/WinningLotto.java:
```diff
@@ -11,8 +11,11 @@
+    private static final int NOT_MATCH = 0;
 
     private final Lotto winningLotto;
+    private LottoNumber bonusBall;
```
bonus번호에 대해 LottoNumber를 재사용하셨군요 좋습니다ㅎㅎ
WinningLotto 객체에 당첨번호, 보너스번호를 같이 관리하는것도 역할에 맞게
잘 부여된것 같네요

limwoobin commented on src/main/java/study/lotto/domain/WinningLotto.java:
```diff
@@ -21,28 +24,51 @@ public WinningLotto(String winningNumbers) {
         return Arrays.stream(winningNumbersSplited)
                 .map((str) -> {
-                    return LottoNumber.of(NumberUtil.convertToPositiveIntNotContainsZero(str.trim()));
+                    return LottoNumber.of(
+                            NumberUtil.convertToPositiveIntNotContainsZero(str.trim())
+                    );
```
로또번호에 대한 validate를 NumberUtil객체를 통해 하고있네요요
이런 에외처리방식보다는 LottoNumber안에서 예외를 처리하는것이 더 좋을것같아요
다른 클라이언트가 해당 객체를 위와 같은 validator없이 바로 사용하는 경우엔 예외를 처리할 수 있습니다.

+ 물론 이 LottoNumber안에선 이미 1~45의 로또번호를 생성해서 검증하기에 예외처리는 되고있겠네요

jisu1211 replied:
제안해주신대로 객체 안에서 명확하게 validation 체크할 수 있도록 고민해보겠습니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/WinningLotto.java:
```diff
@@ -21,28 +24,51 @@ public WinningLotto(String winningNumbers) {
+    public void addBonusBall(int bonusBall) {
+        LottoNumber bonusBallConverted = LottoNumber.of(bonusBall);
+
+        if(!winningLotto.contains(bonusBallConverted)) {
+            this.bonusBall = bonusBallConverted;
+            return;
+        }
+
+        throw new IllegalArgumentException(
+                "[ERROR] Bonus ball number must not be contained in the winning numbers.");
```
저는 개발할때 피치못한 경우가 아니면 부정형은 피해서 작업하는 편인데요
위와같은 경우는 
```java
if (winningLotto.contains(bonusBallConverted)) {
        throw new IllegalArgumentException...
}
``` 
이렇게 보너스번호가 당첨번호에 이미 포함되어 있다면 예외를 발생한다...
와 같은 흐름으로 읽히는게 더 좋지 않을까 의견드려봅니다!

jisu1211 replied:
우빈님이 말씀해주신 내용이 좀 더 깔끔해보이네요! 수정할 때 같이 반영하도록 하겠습니다 ㅎㅎ

limwoobin commented on src/test/java/study/lotto/domain/WinningLottoTest.java:
```diff
@@ -69,4 +73,23 @@ class WinningLottoTest {
         assertEquals(1, tempWinningLotto.matchNumber(LottoNumber.of(num)));
     }
+
+    @ParameterizedTest
+    @ValueSource(ints = { 1, 2, 3, 4, 5, 6 })
+    void addBonusBall_winningNumbers에_포함된_숫자가_보너스볼로_입력되는_경우(int num) {
+        assertThatThrownBy(() -> {
+            tempWinningLotto.addBonusBall(num);
+        }).isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("[ERROR] Bonus ball number must not be contained " +
+                        "in the winning numbers.");
+    }
+
+    @ParameterizedTest
+    @CsvSource(value = { "7:false", "11:true", "9:false", "23:true", "45:true" }, delimiter = ':')
+    void isMatchBonusBall_Lotto가_가지고_있는_숫자_중_bonusBall의_포함_여부(int bonusBall, boolean expected) {
+        Lotto lotto = new Lotto(Arrays.asList(1, 2, 3, 11, 23, 45));
+        tempWinningLotto.addBonusBall(bonusBall);
+
+        assertEquals(expected, tempWinningLotto.isMatchBonusBall(lotto));
+    }
```
보너스볼 추가에 따른 테스트코드도 꼼꼼히 작성해주셨네요~

limwoobin reviewed (changes requested):
안녕하세요 지수님 4단계 미션 굉장히 잘 수행해주셨습니다ㅎㅎ
피드백 반영 또한 굉장히 잘 해주셔서 코드읽기가 훨씬 수월했습니다.

이대로 머지하기엔 고민해보면 좋을 부분이 있어 코멘트 드렸습니다
확인 후 다시 요청부탁드립니다!

+ 4단계 미션내용이 커밋이 한번에 뭉쳐서 잘못올라온걸로 보여집니다.
다음부턴 기능별 작업단위별로 커밋로그를 나눠주시면 더 좋을것같아요!

jisu1211 committed:
fix: 4단계 피드백
- Exception message를 enum 클래스로 관리

jisu1211 committed:
refactor: 4단계 피드백 반영
- LottoNumber를 컬렉션으로 캐싱해두는 구조 -> CacheLottoNumbers로 클래스 분리

test: LottoNumberTest -> CacheLottoNumbersTest로 변경

jisu1211 committed:
refactor: 4단계 피드백 반영
- LottoNumber 클래스 패키지 이동, LottoNumberTest 삭제

jisu1211 committed:
refactor: 유틸성 클래스에 private로 기본생성자 추가

jisu1211 committed:
refactor: 4단계 피드백 반영
- 적절한 메소드 이름으로 변경

jisu1211 committed:
refactor: 4단계 피드백 반영
- Exception Message enum 사용하도록 변경
- Order와 WinningLotto 안에서 Exception 제어

jisu1211 committed:
test: 4단계 피드백 반영
- 코드 변경에 따른 테스트 코드 수정

jisu1211 committed:
refactor: 4단계 피드백 반영
- 같은 이름을 가진 메소드 수정

test: CacheLottoNumbers.ofString 테스트 코드 추가

jisu1211 commented:
피드백 주신 부분들 수정 완료했습니다~!

```
LottoNumber도 1부터 45까지 캐싱을 해주셨군요
만약 그렇다면 이 lottoNumbers를 갖고있는 주체가 LottoNumber가 맞을까요??
```
-> 이렇게 피드백 주신 부분 CacheLottoNumbers 라는 객체로 분리했습니다!

```
예외에 대해 스트링으로 바로 리턴하는 부분이 상당히 보이네요 ㅎㅎ
이런 메시지는 별도의 message constant 객체를 이용하거나 enum을 이용하면
보다 깔끔하고 관리하기 쉽게 사용할 수 있습니다!
```
-> 이 부분은 LottoExceptionCode, NumberExceptionCode  enum 클래스로 추출했습니다~

```
로또번호에 대한 validate를 NumberUtil객체를 통해 하고있네요요
이런 에외처리방식보다는 LottoNumber안에서 예외를 처리하는것이 더 좋을것같아요
다른 클라이언트가 해당 객체를 위와 같은 validator없이 바로 사용하는 경우엔 예외를 처리할 수 있습니다.

물론 이 LottoNumber안에선 이미 1~45의 로또번호를 생성해서 검증하기에 예외처리는 되고있겠네요
```
-> 이 부분은 우빈님이 말씀하신 의도대로 된 것인지는 잘 모르겠으나, 일단 Exception을 catch하여 
error message를 LottoExceptionCode의 에러 메시지로 동일하게 보내도록 수정해봤습니다~
만약 다른 의도셨다면 말씀 부탁드려요 ㅎㅎ

limwoobin commented on src/main/java/study/message/LottoExceptionCode.java:
```diff
@@ -0,0 +1,20 @@
+
+    public String getMessage() {
+        return title + message;
+    }
```
예외 메시지를 한곳에서 관리할 수 있게 분리해주셨군요 좋습니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/number/CacheLottoNumbers.java:
```diff
@@ -0,0 +1,38 @@
+import java.util.Objects;
+import java.util.stream.IntStream;
+
+public class CacheLottoNumbers {
```
클래스에 대한 분리 정말 좋네요 ㅎㅎ
그런데 `Cache...` 라는 네이밍보다는 좀 더 명확한 네이밍은 어떨까요??
로또넘버 객체를 생성 혹은 제공하는거니 ` LottoGenerator` 같은 이름도 좋아보입니다

limwoobin commented on src/main/java/study/lotto/domain/number/LottoNumber.java:
```diff
@@ -0,0 +1,35 @@
+
+    LottoNumber(int lottoNumber) {
+        this.lottoNumber = lottoNumber;
+    }
```
사용되는 scope에 따라 접근제어자를 설정해주셨군요 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/Store.java:
```diff
@@ -28,16 +31,20 @@ private static List<Integer> getRangeOfLottoNumbers() {
+                .stream()
+                .map(CacheLottoNumbers::of)
                 .collect(Collectors.toSet());
     }
```
메소드명을 보니 두가지 일을 하고있는걸로 보여지네요!
클린코드에는 `하나의 메소드는 하나의 기능만 수행한다` 라는 내용이이 있습니다
분리하거나 shuffle이라는 이름은 굳이 기재안해도 될것같아요!
목적은 새로운 로또번호를 리턴하는것이지 로또를 섞는게 아니어서요 ㅎㅎ

https://tecoble.techcourse.co.kr/post/2020-05-10-single-job-method/

jisu1211 replied:
안 그래도 이 부분을 어떻게 작성하는게 맞을지 고민됐었는데 좋은 피드백 감사드려요 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/Order.java:
```diff
@@ -13,11 +14,20 @@ public Order(String totalAmount) {
+            throw new IllegalArgumentException(LottoExceptionCode.INSUFFICIENT_FUNDS.getMessage());
         }
-        throw new IllegalArgumentException("[ERROR] You must purchase at least one lotto.");
     }
```
피드백반영 좋습니다 ㅎㅎ

limwoobin commented on src/main/java/study/lotto/domain/number/CacheLottoNumbers.java:
```diff
@@ -0,0 +1,38 @@
+                .forEach((num) -> cacheLottoNumbers.put(num, new LottoNumber(num)));
+    }
+
+    public static LottoNumber ofString(String number) {
```
메소드명은 굳이 안하셔도될것같아요
하더라도 리턴타입을 명시하는 toLottoNumber 는 어떨까요?
정적팩터리의 네이밍에 대해서도 한번 참고해보시면 좋을것같아요!

https://velog.io/@saint6839/%EC%A0%95%EC%A0%81-%ED%8C%A9%ED%86%A0%EB%A6%AC-%EB%A9%94%EC%84%9C%EB%93%9C-%EB%84%A4%EC%9D%B4%EB%B0%8D-%EB%B0%A9%EC%8B%9D

jisu1211 replied:
보내주신 링크 참고해서 수정하도록 하겠습니다 ㅎㅎ

limwoobin reviewed (approved):
지수님 피드백반영 잘해주셨습니다 ㅎㅎ
굉장히 꼼꼼하게 확인해주셨네요

소소한 코멘트 몇개 남겨드렸습니다. 다음미션 진행하시면서 같이 확인부탁드립니다!
이번단계는 여기서 merge 하겠습니다 다음 단계도 화이팅입니다!

limwoobin commented:
> 피드백 주신 부분들 수정 완료했습니다~!
> 
> ```
> LottoNumber도 1부터 45까지 캐싱을 해주셨군요
> 만약 그렇다면 이 lottoNumbers를 갖고있는 주체가 LottoNumber가 맞을까요??
> ```
> 
> -> 이렇게 피드백 주신 부분 CacheLottoNumbers 라는 객체로 분리했습니다!
> 
> ```
> 예외에 대해 스트링으로 바로 리턴하는 부분이 상당히 보이네요 ㅎㅎ
> 이런 메시지는 별도의 message constant 객체를 이용하거나 enum을 이용하면
> 보다 깔끔하고 관리하기 쉽게 사용할 수 있습니다!
> ```
> 
> -> 이 부분은 LottoExceptionCode, NumberExceptionCode enum 클래스로 추출했습니다~
> 
> ```
> 로또번호에 대한 validate를 NumberUtil객체를 통해 하고있네요요
> 이런 에외처리방식보다는 LottoNumber안에서 예외를 처리하는것이 더 좋을것같아요
> 다른 클라이언트가 해당 객체를 위와 같은 validator없이 바로 사용하는 경우엔 예외를 처리할 수 있습니다.
> 
> 물론 이 LottoNumber안에선 이미 1~45의 로또번호를 생성해서 검증하기에 예외처리는 되고있겠네요
> ```
> 
> -> 이 부분은 우빈님이 말씀하신 의도대로 된 것인지는 잘 모르겠으나, 일단 Exception을 catch하여 error message를 LottoExceptionCode의 에러 메시지로 동일하게 보내도록 수정해봤습니다~ 만약 다른 의도셨다면 말씀 부탁드려요 ㅎㅎ

마지막 코멘트의 의도는 예외처리를 해당 객체안에서 직접 처리하도록 제안드린거였습니다!
말씀하신것처럼 예외처리를 `CacheLottoNumbers`에서 하고있는걸로 보이네요!
